[workspaces] assertion triggered when processing build requirements

Hello,

This one is quite hard to reproduce minimally, I'll try my best to help debug it.
Here is the stacktrace I get when running `conan workspace install`:

>Traceback (most recent call last):
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/command.py", line 1859, in run
    method(args[0][1:])
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/command.py", line 1661, in workspace
    install_folder=args.install_folder)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/conan_api.py", line 78, in wrapper
    return f(*args, **kwargs)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/conan_api.py", line 480, in workspace_install
    False, update, remotes, recorder)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/graph/graph_manager.py", line 182, in load_graph
    graph_lock=graph_lock)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/graph/graph_manager.py", line 284, in _load_graph
    apply_build_requires=apply_build_requires)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/graph/graph_manager.py", line 255, in _recurse_build_requires
    processed_profile, graph_lock)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/graph/graph_manager.py", line 251, in _recurse_build_requires
    processed_profile, graph_lock)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/graph/graph_builder.py", line 65, in extend_build_requires
    remotes, processed_profile, new_reqs, new_options, graph_lock)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/graph/graph_builder.py", line 178, in _handle_require
    update, remotes, processed_profile, graph_lock)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/graph/graph_builder.py", line 132, in _load_deps
    remotes, processed_profile, new_reqs, new_options, graph_lock)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/graph/graph_builder.py", line 200, in _handle_require
    dep_graph.add_edge(node, previous, require)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/graph/graph.py", line 213, in add_edge
    assert src in self.nodes and dst in self.nodes
AssertionError

Here is the conanfile.py causing trouble:

```python
from conans import python_requires


base = python_requires("theo_conanfile/dev")

class BugConan(base.get_conanfile()):
    name = "bug"

    def build_requirements(self):
        if self.should_build_tests:
            # commenting the following line avoids the assert
            self.build_requires("functional-helpers/%s@theo/testing" % self.version)
        # The following adds 3 build requirements, but functional-helpers is not one of them
        super().build_requirements()

    def requirements(self):
        self.requires("log/%s@theo/testing" % self.version)
        self.requires("errors/%s@theo/testing" % self.version)

        self.requires("fmt/5.3.0@bincrafters/stable")
        self.requires("jsonformoderncpp/3.4.0@theo/testing")
        self.requires("optional-lite/3.1.1@theo/testing")
```

I'm using Conan 1.8.1, with revisions enabled.

To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [ ] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

Thanks to @theodelrieu I have a minimum reproducible example with the error: https://github.com/jgsogo/5612-ws-issue/tree/theo-min

Modifying a bit the Conan code, we can get more information for the error:

```bash
    assert dst in self.nodes, "Cannot find dst='{}' in nodes (src='{}')".format(dst, src)
AssertionError: Cannot find dst='fmt/5.3.0@bincrafters/stable' in nodes (src='theo_test-helpers/dev@theo/testing')

ERROR: Cannot find dst='fmt/5.3.0@bincrafters/stable' in nodes (src='theo_test-helpers/dev@theo/testing')
```

I'll add a test reproducing the issue.

--- a/conans/client/graph/graph_manager.py
+++ b/conans/client/graph/graph_manager.py
@@ -214,6 +214,12 @@ class GraphManager(object):
     def _recurse_build_requires(self, graph, subgraph, builder, binaries_analyzer, check_updates,
                                 update, build_mode, remotes, profile_build_requires, recorder,
                                 processed_profile, graph_lock, apply_build_requires=True):
+        """
+        :param graph: This is the full dependency graph with all nodes from all recursions
+        :param subgraph: A partial graph of the nodes that need to be evaluated and expanded
+            at this recursion. Only the nodes belonging to this subgraph will get their package_id
+            computed, and they will resolve build_requires if they need to be built from sources
+        """
 
         binaries_analyzer.evaluate_graph(subgraph, build_mode, update, remotes)
         if not apply_build_requires:

--- a/conans/client/graph/graph_manager.py
+++ b/conans/client/graph/graph_manager.py
@@ -211,15 +211,15 @@ class GraphManager(object):
 
         return conanfile.build_requires
 
-    def _recurse_build_requires(self, graph, builder, binaries_analyzer, check_updates, update,
-                                build_mode, remotes, profile_build_requires, recorder,
+    def _recurse_build_requires(self, graph, subgraph, builder, binaries_analyzer, check_updates,
+                                update, build_mode, remotes, profile_build_requires, recorder,
                                 processed_profile, graph_lock, apply_build_requires=True):
 
-        binaries_analyzer.evaluate_graph(graph, build_mode, update, remotes)
+        binaries_analyzer.evaluate_graph(subgraph, build_mode, update, remotes)
         if not apply_build_requires:
             return
 
-        for node in graph.ordered_iterate():
+        for node in subgraph.ordered_iterate():
             # Virtual conanfiles doesn't have output, but conanfile.py and conanfile.txt do
             # FIXME: To be improved and build a explicit model for this
             if node.recipe == RECIPE_VIRTUAL:
@@ -249,8 +249,8 @@ class GraphManager(object):
                                                          package_build_requires.values(),
                                                          check_updates, update, remotes,
                                                          processed_profile, graph_lock)
-                self._recurse_build_requires(subgraph, builder, binaries_analyzer, check_updates,
-                                             update, build_mode,
+                self._recurse_build_requires(graph, subgraph, builder, binaries_analyzer,
+                                             check_updates, update, build_mode,
                                              remotes, profile_build_requires, recorder,
                                              processed_profile, graph_lock)
                 graph.nodes.update(subgraph.nodes)
@@ -259,8 +259,8 @@ class GraphManager(object):
                 subgraph = builder.extend_build_requires(graph, node, new_profile_build_requires,
                                                          check_updates, update, remotes,
                                                          processed_profile, graph_lock)
-                self._recurse_build_requires(subgraph, builder, binaries_analyzer, check_updates,
-                                             update, build_mode,
+                self._recurse_build_requires(graph, subgraph, builder, binaries_analyzer,
+                                             check_updates, update, build_mode,
                                              remotes, {}, recorder,
                                              processed_profile, graph_lock)
                 graph.nodes.update(subgraph.nodes)
@@ -277,7 +277,7 @@ class GraphManager(object):
         binaries_analyzer = GraphBinariesAnalyzer(self._cache, self._output,
                                                   self._remote_manager)
 
-        self._recurse_build_requires(graph, builder, binaries_analyzer, check_updates, update,
+        self._recurse_build_requires(graph, graph, builder, binaries_analyzer, check_updates, update,
                                      build_mode, remotes,
                                      profile_build_requires, recorder, processed_profile,
                                      graph_lock,

--- a/conans/test/functional/build_requires/build_requires_test.py
+++ b/conans/test/functional/build_requires/build_requires_test.py
@@ -1,10 +1,12 @@
 import os
+import textwrap
 import unittest
 
 from parameterized.parameterized import parameterized
 
+from conans.model.ref import ConanFileReference
 from conans.paths import CONANFILE
-from conans.test.utils.tools import TestClient
+from conans.test.utils.tools import TestClient, GenConanfile
 from conans.util.files import load
 from conans.test.utils.conanfile import TestConanFile
 
@@ -379,3 +381,40 @@ class package(ConanFile):
         client.save({"conanfile.py": consumer})
         client.run("install . --build=missing -o Pkg:someoption=3")
         self.assertIn("first/0.0.0@lasote/stable: Coverage: True", client.out)
+
+    def failed_assert_test(self):
+        # https://github.com/conan-io/conan/issues/5685
+        client = TestClient()
+        client.save({"conanfile.py": GenConanfile()})
+        client.run("export . common/1.0@test/test")
+
+        req = textwrap.dedent("""
+            from conans import ConanFile
+            class BuildReqConan(ConanFile):
+                requires = "common/1.0@test/test"
+            """)
+        client.save({"conanfile.py": req})
+        client.run("export . req/1.0@test/test")
+        client.run("export . build_req/1.0@test/test")
+
+        build_req_req = textwrap.dedent("""
+            from conans import ConanFile
+            class BuildReqConan(ConanFile):
+                requires = "common/1.0@test/test"
+                build_requires = "build_req/1.0@test/test"
+        """)
+        client.save({"conanfile.py": build_req_req})
+        client.run("export . build_req_req/1.0@test/test")
+
+        consumer = textwrap.dedent("""
+                    [requires]
+                    req/1.0@test/test
+                    [build_requires]
+                    build_req_req/1.0@test/test
+                """)
+        client.save({"conanfile.txt": consumer}, clean_first=True)
+        client.run("install . --build=missing")
+        # This used to assert and trace, now it works
+        self.assertIn("conanfile.txt: Applying build-requirement: build_req_req/1.0@test/test",
+                      client.out)
+

--- a/conans/test/functional/workspace/graph_test.py
+++ b/conans/test/functional/workspace/graph_test.py
@@ -1,0 +1,55 @@
+# -*- coding: utf-8 -*-
+import textwrap
+import unittest
+
+from jinja2 import Template
+
+from conans.test.utils.tools import TestClient
+
+
+class WorkspaceTest(unittest.TestCase):
+    conanfile = Template(textwrap.dedent("""
+        from conans import ConanFile
+
+        class Library(ConanFile):
+            name = "{{ name }}"
+            version = "{{ version }}"
+            {% if br %}build_requires = "{{ br }}"{% endif %}
+            {% if req %}requires = "{{ req }}"{% endif %}
+    """))
+
+    def test_graph_failure(self):
+        # https://github.com/conan-io/conan/issues/5612
+        client = TestClient()
+
+        client.save({"fmt/conanfile.py": self.conanfile.render(name="fmt", version="version")})
+
+        client.save({"libA/conanfile.py": self.conanfile.render(name="libA", version="version",
+                                                                req="fmt/version@user/channel")})
+        client.save({"libB/conanfile.py": self.conanfile.render(name="libB", version="version",
+                                                                req="fmt/version@user/channel",
+                                                                br="libA/version@user/channel")})
+        client.save({"libC/conanfile.py": self.conanfile.render(name="libC", version="version",
+                                                                req="fmt/version@user/channel",
+                                                                br="libB/version@user/channel")})
+
+        client.run("create fmt fmt/version@user/channel")
+
+        client.run("create libA libA/version@user/channel")
+        client.run("create libB libB/version@user/channel")
+        client.run("create libC libC/version@user/channel")
+
+        conanws = textwrap.dedent("""
+                    editables:
+                      libA/version@user/channel:
+                        path: libA
+                      libB/version@user/channel:
+                        path: libB
+                      libC/version@user/channel:
+                        path: libC
+        
+                    workspace_generator: cmake
+                    root: libC/version@user/channel
+                """)
+        client.save({"conanws.yml": conanws})
+        client.run("workspace install .")

